### PR TITLE
VACMS-1651 site alert no cache

### PR DESCRIPTION
**Pull request by indytechcook**
## Description

See #1651 

## Testing done

* Set the env variable `VA_GOV_IN_MAINTENANCE_MODE` to `TRUE` and watched the site alert come back the same.
* Removed the env variable and it went away correctly.

---
This change is [Reviewable](https://reviewable.io/reviews/department-of-veterans-affairs/va.gov-cms/1954)

### CODE REVIEW

File: docroot/sites/default/settings.php [hunk 1]
```diff
@@ -158,6 +158,16 @@
 $settings['slack_webhook_url'] = getenv('CMS_VAGOV_SLACK_WEBHOOK_URL') ?: FALSE;
 $config['slack.settings']['slack_webhook_url'] = $settings['slack_webhook_url'];
 
+// Site Alert settings
+$settings['va_gov_site_alert_html'] = <<<EOF
```
Comment (indytechcook) [inline docroot/sites/default/settings.php:162]: @swirtSJW  Another solution would be a html file which is updated when site alerts are created via an update hook.

File: docroot/sites/default/settings.php [hunk 2]
```diff
@@ -183,5 +198,19 @@
...
+// If in maintenance mode then process site alert
+if (!empty(getenv('VA_GOV_IN_MAINTENANCE_MODE'))) {
+  $html = $settings['va_gov_site_alert_html'] ?? '';
+  $site_alert_paths = $settings['va_gov_site_alert_paths'] ?? ['ajax/site_alert'];
+  $current_path = $GLOBALS['request']->getPathInfo();
+  $trimmed_path = ltrim($current_path, '/');
+
+  if (in_array($trimmed_path, $site_alert_paths)) {
+    // Throw a Custom except to skip the rest of the bootstrap process and avoid cache.
+    include_once $app_root . '/modules/custom/va_gov_backend/src/Exception/SiteAlertException.php';
```
Comment (indytechcook) [inline docroot/sites/default/settings.php:210]: I wasn't here exactly where to put this file.

File: composer.lock [hunk 3]
```diff
@@ -9728,6 +9740,62 @@
             },
             "time": "2020-04-02T17:11:23+00:00"
         },
+        {
+            "name": "drupal/skip_all_caches",
```
Comment (swirtSJW) [inline composer.lock:9744]: Is this still needed, or leftover from previous?

File: docroot/sites/default/settings.php (hunk 1 above)
Comment (swirtSJW) [reply docroot/sites/default/settings.php:162]: How would the update hook run more than once?

Comment (indytechcook) [reply docroot/sites/default/settings.php:162]: I meant an entity update hook

File: docroot/sites/default/settings/deploy-settings.deploy.php [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<?php
```
Comment (swirtSJW) [inline docroot/sites/default/settings/deploy-settings.deploy.php:1]: Suggest changing this to `deploy-settings.active.php` and togle to `deploy-settings.inactive.php`

File: docroot/sites/default/settings.php [hunk 5]
```diff
@@ -184,5 +200,20 @@
   include $app_root . '/' . $site_path . '/settings/settings.local.php';
 }
 
+// The VA_GOV_IN_MAINTENANCE_MODE is set in settings.deploy.active.php.
```
Comment (indytechcook) [inline docroot/sites/default/settings.php:203]:
```suggestion
// The VA_GOV_IN_DEPLOY_MODE is set in settings.deploy.active.php.
```

File: docroot/modules/custom/va_gov_backend/src/Exception/SuccessHTTPException.php [hunk 6]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Drupal\va_gov_backend\Exception;
+
+use Symfony\Component\HttpKernel\Exception\HttpException;
+
+/**
+ * An excpetion which returns a 200.
```
Comment (ElijahLynn) [inline docroot/modules/custom/va_gov_backend/src/Exception/SuccessHTTPException.php:8]:
```suggestion
 * An exception which returns a 200.
```

File: docroot/modules/custom/va_gov_backend/src/Exception/SuccessHTTPException.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * This allows for stopping a HTTP request and returning a 200 and static HTML.
```
Comment (ElijahLynn) [inline docroot/modules/custom/va_gov_backend/src/Exception/SuccessHTTPException.php:10]:
```suggestion
 * This allows for stopping a HTTP request from bootstrapping Drupal, then returning a 200 with static HTML.
```

File: docroot/sites/default/settings.php [hunk 8]
```diff
@@ -159,6 +159,17 @@
 $settings['slack_webhook_url'] = getenv('CMS_VAGOV_SLACK_WEBHOOK_URL') ?: FALSE;
 $config['slack.settings']['slack_webhook_url'] = $settings['slack_webhook_url'];
 
+// Site Alert settings.
+// If you update this text also update the site alert in the devops repo.
+$settings['va_gov_bybass_message_html'] = <<<EOF
```
Comment (ElijahLynn) [inline docroot/sites/default/settings.php:164]:
```suggestion
// Site Alert settings.
// If you update this text also update the site alert in the devops repo.
// @TODO: Move this to a separate file and include it. 
$settings['va_gov_bybass_message_html'] = <<<EOF
```

File: docroot/sites/default/settings.php [hunk 9]
```diff
@@ -169,6 +180,11 @@
   include $app_root . '/' . $site_path . '/settings/settings.fast_404.php';
 }
 
+// Deploy Settings
+if (file_exists($app_root . '/' . $site_path . '/settings/settings.deploy.active.php')) {
```
Comment (ElijahLynn) [inline docroot/sites/default/settings.php:184]: Rename this one or the one in https://github.com/department-of-veterans-affairs/devops/pull/7433/files#diff-bb36423fc09dbc13ffa8ccc5c56144d0R230.

Comment (indytechcook) [reply docroot/sites/default/settings.php:185]: The other PR was updated. These file names were set after a discussion with @swirtSJW

File: docroot/modules/custom/va_gov_backend/src/Exception/SuccessHTTPException.php [hunk 10]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Drupal\va_gov_backend\Exception;
+
+use Symfony\Component\HttpKernel\Exception\HttpException;
+
+/**
+ * An exception which returns a 200.
+ *
+ * This allows for stopping a HTTP request from bootstrapping Drupal, then returning a 200 with static HTML.
```
Comment (ElijahLynn) [inline docroot/modules/custom/va_gov_backend/src/Exception/SuccessHTTPException.php:10]:
```suggestion
 * This allows for stopping a HTTP request from bootstrapping Drupal, then returning a 200 
 * with static HTML.
```

Comment (ElijahLynn) [inline docroot/modules/custom/va_gov_backend/src/Exception/SuccessHTTPException.php:10]:
```suggestion
 * This allows for stopping a HTTP request from bootstrapping Drupal, 
 * then returning a 200 with static HTML.
```

**Comment by ElijahLynn** [comment]
Neil and I just chatted about this on a call and we discussed the fact that this is needed was because maintenance mode broke things with an infinite redirect when users logged in, thus causing a accidental DDoS with logged in users. @swirtSJW, I see your patch for Redirect After Login module here https://www.drupal.org/project/redirect_after_login/issues/3133579 and that seemed like it would work but then I realized we are not using that module right now. 

I did verify that maintenance mode is still causing infinite redirects for users. 

I think we should hold off on this until we figure out that issue because using maintenance mode should help us here, yah?

File: docroot/sites/default/settings.php [hunk 11]
```diff
@@ -184,5 +201,20 @@
   include $app_root . '/' . $site_path . '/settings/settings.local.php';
 }
 
+// The VA_GOV_IN_DEPLOY_MODE is set in settings.deploy.active.php.
+// Ths file is copied from settings.deploy.inactive.php. by ansible during deploys.
+if (!empty(getenv('VA_GOV_IN_DEPLOY_MODE'))) {
+  $html = $settings['va_gov_bybass_message_html'] ?? '';
```
Comment (indytechcook) [inline docroot/sites/default/settings.php:207]:
```suggestion
  $html = $settings['va_gov_site_alert_bybass_message_html'] ?? '';
```

File: docroot/sites/default/settings.php [hunk 12]
```diff
@@ -159,6 +159,18 @@
 $settings['slack_webhook_url'] = getenv('CMS_VAGOV_SLACK_WEBHOOK_URL') ?: FALSE;
 $config['slack.settings']['slack_webhook_url'] = $settings['slack_webhook_url'];
 
+// Site Alert settings.
+// If you update this text also update the site alert in the devops repo.
+// @TODO: Move this to a separate file and include it. 
+$settings['va_gov_bybass_message_html'] = <<<EOF
```
Comment (indytechcook) [inline docroot/sites/default/settings.php:165]:
```suggestion
$settings['va_gov_site_alert_bybass_message_html'] = <<<EOF
```

**Comment by indytechcook** [comment]
Closing this in favor of https://github.com/department-of-veterans-affairs/va.gov-cms/pull/2878

**Pull request closed without merging**